extract/extract/extractors/amazon.py

```diff
@@ -0,0 +1,298 @@
+    """
+    if any(targets):
+        result = [target for target in targets if target][0]
+        return parse(result)
+
+    return None
+
+
+def _get_color(soup: BeautifulSoup) -> Optional[str]:
+    """
+    Helper function that extracts the product's color.
+
+    Args:
+        soup (BeautifulSoup): Parsed HTML
+
+    Returns:
+        Optional[str]: `str` object with the color name. If nothing was found `None` is returned.
+    """
+    targets = [
+        soup.find("span", {"class": "selection"}),
+        soup.find("tr", {"class": re.compile("po-color")}),
+    ]
+
+    def parse_color(color: BeautifulSoup) -> Optional[str]:
+        if color.name == "span":
+            return color.text.strip()
+        if color.name == "tr":
+            return color.find_all("span")[1].text
+        return None
+
+    return _handle_parse(targets, parse_color)
+
+
+def _get_image_urls(soup: BeautifulSoup) -> Optional[list[str]]:
+    """
+    Helper function that extracts the product's image urls.
+
+    Args:
+        soup (BeautifulSoup): Parsed HTML
+
+    Returns:
+        Optional[list[str]]: `list` object containing `str` objects representing the image urls.
+            If nothing was found `None` or empty list is returned.
+    """
+    targets = [
+        soup.find("div", {"id": "altImages"}).find_all("img"),
+    ]
+
+    def parse_image_urls(images: list[BeautifulSoup]) -> list[str]:
+        image_urls = [
+            image["src"]
+            for image in images
+            if not image["src"].endswith(".gif")
+            and "play-button-overlay" not in image["src"]
+            and "play-icon-overlay" not in image["src"]
+            and "360_icon" not in image["src"]
+        ]
+        return image_urls
+
+    return _handle_parse(targets, parse_image_urls)
+
+
+def _get_sizes(soup: BeautifulSoup) -> Optional[str]:
+    """
+    Helper function that extracts the product's sizes.
+
+    Args:
+        soup (BeautifulSoup): Parsed HTML
+
+    Returns:
+        Optional[str]: `str` object containing all sizes. If nothing was found `None`.
+    """
+    targets = [
+        soup.find_all("span", {"class", "a-size-base swatch-title-text-display swatch-title-text"})[
+            1:
+        ],
+        soup.find_all("option", id=re.compile("size_name"))[1:],
+    ]
+
+    def parse_sizes(sizes: list[BeautifulSoup]) -> str:
+        sizes = [size.text.strip() for size in sizes if size.text.strip()]
+        return ", ".join(sizes)
+
+    return _handle_parse(targets, parse_sizes)
+
+
+def _get_price(parsed_page: ParsedPage) -> Optional[float]:
+    """
+    Helper function that extracts the product's price.
+
+    Args:
+        parsed_page (ParsedPage): Intermediate representation of `ScrapedPage` domain object
+
+    Returns:
+        Optional[float]: `float` object if a price is given, else `None`.
+    """
+    targets = [
+        parsed_page.scraped_page.meta_information["amazon_price"],
```

Comment:
..finally here: `parsed_page.scraped_page.meta_information.get("request_meta_information", {}).get("price", None)`
It is better to use default return value that can be handled later on (e.g. empty dict). If not, it could raise errors.

For extractors, we should always try to construct `Product` object. Here, we could set `None` as price. Of course, this is not valid and `Product(...)` will raise an exception. However, it is nicely structured and contains valuable debug information about why Product is not valid.

Reply:
Changed in [74f677d](https://github.com/calgo-lab/green-db/pull/65/commits/74f677d372ac18391ade43d100136d7839f61e54)